File: backend/src/models/Lexicon.ts
import mongoose, { Schema, Document } from 'mongoose';

// A per-language, library-wide dictionary applied at read time (TTS path only).
// Currently holds acronym/version expansions (e.g. NVI -> Nova Versão Internacional).
export interface IAcronym {
  term: string;   // matched whole-word, case-sensitive (e.g. "NVI")
  say: string;    // spoken expansion
}

export interface ILexicon extends Document {
  language: string;          // ISO 639-1 (e.g. 'en', 'pt')
  acronyms: IAcronym[];
  updatedAt: Date;
}

const AcronymSchema = new Schema<IAcronym>(
  { term: { type: String, required: true }, say: { type: String, required: true } },
  { _id: false },
);

const LexiconSchema = new Schema<ILexicon>(
  {
    language: { type: String, required: true, unique: true, index: true },
    acronyms: { type: [AcronymSchema], default: [] },
  },
  { timestamps: true },
);

export const Lexicon = mongoose.model<ILexicon>('Lexicon', LexiconSchema);

// Default Bible-version acronyms seeded per language. Users can edit via the API.
const DEFAULT_ACRONYMS: Record<string, IAcronym[]> = {
  en: [
    { term: 'KJV', say: 'King James Version' },
    { term: 'NKJV', say: 'New King James Version' },
    { term: 'NIV', say: 'New International Version' },
    { term: 'ESV', say: 'English Standard Version' },
    { term: 'NLT', say: 'New Living Translation' },
    { term: 'NASB', say: 'New American Standard Bible' },
    { term: 'NRSV', say: 'New Revised Standard Version' },
    { term: 'CSB', say: 'Christian Standard Bible' },
    { term: 'ASV', say: 'American Standard Version' },
    { term: 'AMP', say: 'Amplified Bible' },
    { term: 'e.g.', say: 'for example' },
    { term: 'i.e.', say: 'that is' },
    { term: 'cf.', say: 'compare' },
    { term: '=', say: 'equals' },
  ],
  pt: [
    { term: 'NVI', say: 'Nova Versão Internacional' },
    { term: 'ARA', say: 'Almeida Revista e Atualizada' },
    { term: 'ARC', say: 'Almeida Revista e Corrigida' },
    { term: 'ACF', say: 'Almeida Corrigida Fiel' },
    { term: 'NTLH', say: 'Nova Tradução na Linguagem de Hoje' },
    { term: 'NAA', say: 'Nova Almeida Atualizada' },
    { term: 'NVT', say: 'Nova Versão Transformadora' },
    { term: 'KJA', say: 'King James Atualizada' },
    { term: 'e.g.', say: 'por exemplo' },
    { term: 'i.e.', say: 'isto é' },
    { term: 'cf.', say: 'confira' },
    { term: '=', say: 'igual a' },
  ],
};

// Seed the per-language doc when missing, and add any default terms absent from
// an existing doc (e.g. defaults shipped after first seed) without clobbering
// user edits.
export async function seedLexicons(): Promise<void> {
  for (const [language, acronyms] of Object.entries(DEFAULT_ACRONYMS)) {
    await Lexicon.updateOne(
      { language },
      { $setOnInsert: { language, acronyms } },
      { upsert: true },
    );
    const doc = await Lexicon.findOne({ language });
    if (!doc) continue;
    const have = new Set(doc.acronyms.map(a => a.term));
    const missing = acronyms.filter(a => !have.has(a.term));
    if (missing.length) {
      doc.acronyms.push(...missing);
      await doc.save();
    }
  }
}

File: backend/internal/model/lexicon.go
package model

import (
	"go.mongodb.org/mongo-driver/v2/bson"
)

type Acronym struct {
	Term string `bson:"term" json:"term"`
	Say  string `bson:"say"  json:"say"`
}

// Lexicon is a per-language, library-wide dictionary applied at read time
// (TTS path only) — acronym/version expansions like NVI → Nova Versão
// Internacional.
type Lexicon struct {
	ID        bson.ObjectID `bson:"_id"       json:"_id"`
	Language  string        `bson:"language"  json:"language"`
	Acronyms  []Acronym     `bson:"acronyms"  json:"acronyms"`
	CreatedAt DateTime      `bson:"createdAt" json:"createdAt"`
	UpdatedAt DateTime      `bson:"updatedAt" json:"updatedAt"`
	V         int32         `bson:"__v"       json:"__v"`
}

// DefaultAcronyms are the Bible-version acronyms seeded per language. Users
// can edit via the API.
var DefaultAcronyms = map[string][]Acronym{
	"en": {
		{Term: "KJV", Say: "King James Version"},
		{Term: "NKJV", Say: "New King James Version"},
		{Term: "NIV", Say: "New International Version"},
		{Term: "ESV", Say: "English Standard Version"},
		{Term: "NLT", Say: "New Living Translation"},
		{Term: "NASB", Say: "New American Standard Bible"},
		{Term: "NRSV", Say: "New Revised Standard Version"},
		{Term: "CSB", Say: "Christian Standard Bible"},
		{Term: "ASV", Say: "American Standard Version"},
		{Term: "AMP", Say: "Amplified Bible"},
		{Term: "e.g.", Say: "for example"},
		{Term: "i.e.", Say: "that is"},
		{Term: "cf.", Say: "compare"},
		{Term: "=", Say: "equals"},
	},
	"pt": {
		{Term: "NVI", Say: "Nova Versão Internacional"},
		{Term: "ARA", Say: "Almeida Revista e Atualizada"},
		{Term: "ARC", Say: "Almeida Revista e Corrigida"},
		{Term: "ACF", Say: "Almeida Corrigida Fiel"},
		{Term: "NTLH", Say: "Nova Tradução na Linguagem de Hoje"},
		{Term: "NAA", Say: "Nova Almeida Atualizada"},
		{Term: "NVT", Say: "Nova Versão Transformadora"},
		{Term: "KJA", Say: "King James Atualizada"},
		{Term: "e.g.", Say: "por exemplo"},
		{Term: "i.e.", Say: "isto é"},
		{Term: "cf.", Say: "confira"},
		{Term: "=", Say: "igual a"},
	},
}
